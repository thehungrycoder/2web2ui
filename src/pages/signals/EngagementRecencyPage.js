import React, { Component, Fragment } from 'react';
import { Link } from 'react-router-dom';
import { Panel, Grid } from '@sparkpost/matchbox';
import Page from './components/SignalsPage';
import BarChart from './components/charts/barchart/BarChart';
import EngagementRecencyActions from './components/actionContent/EngagementRecencyActions';
import TooltipMetric from './components/charts/tooltip/TooltipMetric';
import DateFilter from './components/filters/DateFilter';
import { ENGAGEMENT_RECENCY_COHORTS, ENGAGEMENT_RECENCY_INFO } from './constants/info';
import withEngagementRecencyDetails from './containers/EngagementRecencyDetailsContainer';
import { Loading } from 'src/components';
import Callout from 'src/components/callout';
import OtherChartsHeader from './components/OtherChartsHeader';
import ChartHeader from './components/ChartHeader';
import Legend from './components/charts/legend/Legend';
import { roundToPlaces } from 'src/helpers/units';
import moment from 'moment';
import _ from 'lodash';

import SpamTrapsPreview from './components/previews/SpamTrapsPreview';
import HealthScorePreview from './components/previews/HealthScorePreview';
import cohorts from './constants/cohorts';
import styles from './DetailsPages.module.scss';

export class EngagementRecencyPage extends Component {
  state = {
    selectedDate: null
  }

  componentDidMount() {
    const { selected } = this.props;

    if (selected) {
      this.setState({ selectedDate: selected });
    }
  }

  componentDidUpdate(prevProps) {
    const { data } = this.props;
    const { selectedDate } = this.state;

    const dataSetChanged = prevProps.data !== data;
    let selectedDataByDay = _.find(data, ['date', selectedDate]);

    // Select last date in time series
    if (dataSetChanged && !selectedDataByDay) {
      selectedDataByDay = _.last(data);
      this.setState({ selectedDate: selectedDataByDay.date });
    }
  }

  handleDateSelect = (node) => {
    this.setState({ selectedDate: _.get(node, 'payload.date') });
  }

  getYAxisProps = () => ({
    tickFormatter: (tick) => `${roundToPlaces(tick * 100, 0)}%`,
    domain: [0, 1],
    ticks: [0, 0.25, 0.5, 0.75, 1.0]
  })

  getXAxisProps = () => {
    const { xTicks } = this.props;
    return {
      ticks: xTicks,
      tickFormatter: (tick) => moment(tick).format('M/D')
    };
  }

  getTooltipContent = ({ payload = {}}) => (
    <Fragment>
      {_.keys(cohorts).map((key) => (
        <TooltipMetric
          key={key}
          color={cohorts[key].fill}
          label={cohorts[key].label}
          value={`${roundToPlaces(payload[key] * 100, 1)}%`}
        />
      ))}
    </Fragment>
  )

  renderContent = () => {
    const { data = [], loading, gap, empty, error } = this.props;
    const { selectedDate } = this.state;
    const selectedData = _.find(data, ['date', selectedDate]) || {};
    let chartPanel;

    if (empty) {
      chartPanel = <Callout title='No Data Available'>Insufficient data to populate this chart</Callout>;
    }

    if (error) {
      chartPanel = <Callout title='Unable to Load Data'>{error.message}</Callout>;
    }

    if (loading) {
      chartPanel = (
        <div style={{ height: '220px', position: 'relative' }}>
          <Loading />
        </div>
      );
    }

    return (
      <Grid>
        <Grid.Column sm={12} md={7}>
          <Panel sectioned>
            <ChartHeader
              title='Engagement Recency'
              tooltipContent={ENGAGEMENT_RECENCY_INFO}
            />
            {chartPanel || (
              <div className='LiftTooltip'>
                <BarChart
                  gap={gap}
                  onClick={this.handleDateSelect}
                  selected={selectedDate}
                  timeSeries={data}
                  tooltipContent={this.getTooltipContent}
                  yKeys={_.keys(cohorts).map((key) => ({ key, ...cohorts[key] })).reverse()}
                  yAxisProps={this.getYAxisProps()}
                  xAxisProps={this.getXAxisProps()}
                />
                <Legend
                  items={_.values(cohorts)}
                  tooltipContent={(label) => ENGAGEMENT_RECENCY_COHORTS[label]}
                />
              </div>
            )}
          </Panel>
        </Grid.Column>
        <Grid.Column sm={12} md={5} mdOffset={0}>
          <div className={styles.OffsetCol}>
            {!chartPanel && <EngagementRecencyActions cohorts={selectedData} date={selectedDate} />}
          </div>
        </Grid.Column>
      </Grid>
    );
  }

  render() {
    const { facet, facetId } = this.props;

    return (
      <Page
        breadcrumbAction={{ content: 'Back to Overview', to: '/signals', component: Link }}
        dimensionPrefix='Engagement Recency for'
        facet={facet}
        facetId={facetId}
        primaryArea={<DateFilter />}>
        {this.renderContent()}
        <OtherChartsHeader facet={facet} facetId={facetId} />
        <Grid>
          <Grid.Column xs={12} sm={6}>
            <SpamTrapsPreview />
          </Grid.Column>
          <Grid.Column xs={12} sm={6}>
            <HealthScorePreview />
          </Grid.Column>
        </Grid>
      </Page>
    );
  }
}

export default withEngagementRecencyDetails(EngagementRecencyPage);
