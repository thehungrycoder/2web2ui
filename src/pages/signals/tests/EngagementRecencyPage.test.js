import { shallow } from 'enzyme';
import React from 'react';
import { EngagementRecencyPage } from '../EngagementRecencyPage';

describe('Signals Engagement Recency Page', () => {
  let wrapper;
  let props;
  const data = [
    {
      date: '2017-01-01',
      c_total: 1
    },
    {
      date: '2017-01-02',
      c_total: 10
    }
  ];

  beforeEach(() => {
    props = {
      facetId: 'test.com',
      facet: 'sending-domain',
      data: [{ c_total: 10 }],
      gap: 0.25,
      loading: false,
      empty: false,
      xTicks: [1,2]
    };
    wrapper = shallow(<EngagementRecencyPage {...props}/>);
    wrapper.setProps({ data });
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders loading correctly', () => {
    wrapper.setProps({ loading: true });
    expect(wrapper.find('Panel')).toMatchSnapshot();
  });

  it('renders empty correctly', () => {
    wrapper.setProps({ empty: true });
    expect(wrapper.find('Panel')).toMatchSnapshot();
  });

  it('renders error correctly', () => {
    wrapper.setProps({ error: { message: 'error message' }});
    expect(wrapper).toMatchSnapshot();
  });

  describe('local state', () => {
    it('handles date select', () => {
      wrapper.find('BarChart').simulate('click', { payload: { date: '2017-01-02' }});
      expect(wrapper.find('BarChart').prop('selected')).toEqual('2017-01-02');
      expect(wrapper.find('EngagementRecencyActions')).toMatchSnapshot();
    });

    it('sets selected date on mount if provided one', () => {
      wrapper = shallow(<EngagementRecencyPage {...props} selected='initial-selected'/>);
      expect(wrapper.find('BarChart').prop('selected')).toEqual('initial-selected');
      expect(wrapper.find('EngagementRecencyActions').prop('date')).toEqual('initial-selected');
    });

    it('uses last selected date if selected date is not in data', () => {
      wrapper = shallow(<EngagementRecencyPage {...props} loading={true} selected='initial-selected'/>);
      wrapper.setProps({ data: [1, { date: 'last-date' }], loading: false });
      expect(wrapper.find('BarChart').prop('selected')).toEqual('last-date');
      expect(wrapper.find('EngagementRecencyActions').prop('date')).toEqual('last-date');
    });

    it('does not use last selected date if selected date is in data', () => {
      wrapper = shallow(<EngagementRecencyPage {...props} selected='first-date'/>);
      wrapper.setProps({ data: [{ date: 'first-date' }, { date: 'last-date' }]});
      expect(wrapper.find('BarChart').prop('selected')).toEqual('first-date');
      expect(wrapper.find('EngagementRecencyActions').prop('date')).toEqual('first-date');
    });
  });

  describe('bar chart props', () => {
    it('renders tooltip content', () => {
      const Tooltip = wrapper.find('BarChart').prop('tooltipContent');
      expect(shallow(<Tooltip payload={{
        c_uneng: 0.1,
        c_365d: 0.2,
        c_90d: 0.3,
        c_14d: 0.4,
        c_new: 0.5,
        date: '2018-01-01',
        c_total: 10
      }} />)).toMatchSnapshot();
    });

    it('gets x axis props', () => {
      const axisProps = wrapper.find('BarChart').prop('xAxisProps');
      expect(axisProps).toMatchSnapshot();
      expect(axisProps.tickFormatter('2018-12-05')).toEqual('12/5');
    });

    it('gets y axis props', () => {
      const axisProps = wrapper.find('BarChart').prop('yAxisProps');
      expect(axisProps.tickFormatter(0.2468)).toEqual('25%');
    });
  });
});
