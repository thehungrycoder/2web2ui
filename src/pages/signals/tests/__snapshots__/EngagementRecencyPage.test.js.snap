// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signals Engagement Recency Page bar chart props gets x axis props 1`] = `
Object {
  "tickFormatter": [Function],
  "ticks": Array [
    1,
    2,
  ],
}
`;

exports[`Signals Engagement Recency Page bar chart props renders tooltip content 1`] = `
<Fragment>
  <TooltipMetric
    color="#E4EEF0"
    key="c_uneng"
    label="Never Engaged"
    value="10%"
  />
  <TooltipMetric
    color="#B9E2E9"
    key="c_365d"
    label="Not Recently Engaged"
    value="20%"
  />
  <TooltipMetric
    color="#9AD8E3"
    key="c_90d"
    label="Semi Recently Engaged"
    value="30%"
  />
  <TooltipMetric
    color="#44A8AD"
    key="c_14d"
    label="Recently Engaged"
    value="40%"
  />
  <TooltipMetric
    color="#3F8B93"
    key="c_new"
    label="New"
    value="50%"
  />
</Fragment>
`;

exports[`Signals Engagement Recency Page local state handles date select 1`] = `
<EngagementRecencyActions
  cohorts={
    Object {
      "c_total": 10,
      "date": "2017-01-02",
    }
  }
  date="2017-01-02"
/>
`;

exports[`Signals Engagement Recency Page renders correctly 1`] = `
<SignalsPage
  breadcrumbAction={
    Object {
      "component": [Function],
      "content": "Back to Overview",
      "to": "/signals",
    }
  }
  dimensionPrefix="Engagement Recency for"
  facet="sending-domain"
  facetId="test.com"
  primaryArea={<Connect(DateFilter) />}
>
  <Grid>
    <Grid.Column
      md={7}
      sm={12}
    >
      <Panel
        sectioned={true}
      >
        <ChartHeader
          title="Engagement Recency"
          tooltipContent="
  The share over time of your email that has been sent to recipients who most recently
  opened messages or clicked links during several defined time periods.
"
        />
        <div
          className="LiftTooltip"
        >
          <BarChart
            fill="#B157CE"
            gap={0.25}
            height={250}
            margin={
              Object {
                "bottom": 5,
                "left": 18,
                "right": 0,
                "top": 12,
              }
            }
            onClick={[Function]}
            selected="2017-01-02"
            timeSeries={
              Array [
                Object {
                  "c_total": 1,
                  "date": "2017-01-01",
                },
                Object {
                  "c_total": 10,
                  "date": "2017-01-02",
                },
              ]
            }
            tooltipContent={[Function]}
            width="99%"
            xAxisProps={
              Object {
                "tickFormatter": [Function],
                "ticks": Array [
                  1,
                  2,
                ],
              }
            }
            xKey="date"
            yAxisProps={
              Object {
                "domain": Array [
                  0,
                  1,
                ],
                "tickFormatter": [Function],
                "ticks": Array [
                  0,
                  0.25,
                  0.5,
                  0.75,
                  1,
                ],
              }
            }
            yKey="value"
            yKeys={
              Array [
                Object {
                  "fill": "#3F8B93",
                  "key": "c_new",
                  "label": "New",
                },
                Object {
                  "fill": "#44A8AD",
                  "key": "c_14d",
                  "label": "Recently Engaged",
                },
                Object {
                  "fill": "#9AD8E3",
                  "key": "c_90d",
                  "label": "Semi Recently Engaged",
                },
                Object {
                  "fill": "#B9E2E9",
                  "key": "c_365d",
                  "label": "Not Recently Engaged",
                },
                Object {
                  "fill": "#E4EEF0",
                  "key": "c_uneng",
                  "label": "Never Engaged",
                },
              ]
            }
            yRange={
              Array [
                "auto",
                "auto",
              ]
            }
          />
          <Legend
            items={
              Array [
                Object {
                  "fill": "#E4EEF0",
                  "label": "Never Engaged",
                },
                Object {
                  "fill": "#B9E2E9",
                  "label": "Not Recently Engaged",
                },
                Object {
                  "fill": "#9AD8E3",
                  "label": "Semi Recently Engaged",
                },
                Object {
                  "fill": "#44A8AD",
                  "label": "Recently Engaged",
                },
                Object {
                  "fill": "#3F8B93",
                  "label": "New",
                },
              ]
            }
            tooltipContent={[Function]}
          />
        </div>
      </Panel>
    </Grid.Column>
    <Grid.Column
      md={5}
      mdOffset={0}
      sm={12}
    >
      <div
        className="OffsetCol"
      >
        <EngagementRecencyActions
          cohorts={
            Object {
              "c_total": 10,
              "date": "2017-01-02",
            }
          }
          date="2017-01-02"
        />
      </div>
    </Grid.Column>
  </Grid>
  <OtherChartsHeader
    facet="sending-domain"
    facetId="test.com"
  />
  <Grid>
    <Grid.Column
      sm={6}
      xs={12}
    >
      <withRouter(Connect(WithSpamTrapDetails(SpamTrapsPreview))) />
    </Grid.Column>
    <Grid.Column
      sm={6}
      xs={12}
    >
      <withRouter(Connect(WithHealthScoreDetails(HealthScorePreview))) />
    </Grid.Column>
  </Grid>
</SignalsPage>
`;

exports[`Signals Engagement Recency Page renders empty correctly 1`] = `
<Panel
  sectioned={true}
>
  <ChartHeader
    title="Engagement Recency"
    tooltipContent="
  The share over time of your email that has been sent to recipients who most recently
  opened messages or clicked links during several defined time periods.
"
  />
  <Callout
    title="No Data Available"
  >
    Insufficient data to populate this chart
  </Callout>
</Panel>
`;

exports[`Signals Engagement Recency Page renders error correctly 1`] = `
<SignalsPage
  breadcrumbAction={
    Object {
      "component": [Function],
      "content": "Back to Overview",
      "to": "/signals",
    }
  }
  dimensionPrefix="Engagement Recency for"
  facet="sending-domain"
  facetId="test.com"
  primaryArea={<Connect(DateFilter) />}
>
  <Grid>
    <Grid.Column
      md={7}
      sm={12}
    >
      <Panel
        sectioned={true}
      >
        <ChartHeader
          title="Engagement Recency"
          tooltipContent="
  The share over time of your email that has been sent to recipients who most recently
  opened messages or clicked links during several defined time periods.
"
        />
        <Callout
          title="Unable to Load Data"
        >
          error message
        </Callout>
      </Panel>
    </Grid.Column>
    <Grid.Column
      md={5}
      mdOffset={0}
      sm={12}
    >
      <div
        className="OffsetCol"
      />
    </Grid.Column>
  </Grid>
  <OtherChartsHeader
    facet="sending-domain"
    facetId="test.com"
  />
  <Grid>
    <Grid.Column
      sm={6}
      xs={12}
    >
      <withRouter(Connect(WithSpamTrapDetails(SpamTrapsPreview))) />
    </Grid.Column>
    <Grid.Column
      sm={6}
      xs={12}
    >
      <withRouter(Connect(WithHealthScoreDetails(HealthScorePreview))) />
    </Grid.Column>
  </Grid>
</SignalsPage>
`;

exports[`Signals Engagement Recency Page renders loading correctly 1`] = `
<Panel
  sectioned={true}
>
  <ChartHeader
    title="Engagement Recency"
    tooltipContent="
  The share over time of your email that has been sent to recipients who most recently
  opened messages or clicked links during several defined time periods.
"
  />
  <div
    style={
      Object {
        "height": "220px",
        "position": "relative",
      }
    }
  >
    <Loading />
  </div>
</Panel>
`;
